src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc

```diff
@@ -60,165 +55,95 @@ namespace grpc_core {
 
 namespace {
 
-class AresClientChannelDNSResolver : public Resolver {
+class AresClientChannelDNSResolver : public PollingResolver {
  public:
-  explicit AresClientChannelDNSResolver(ResolverArgs args);
+  AresClientChannelDNSResolver(ResolverArgs args,
+                               const grpc_channel_args* channel_args);
 
-  void StartLocked() override;
+  OrphanablePtr<Orphanable> StartRequest() override;
 
-  void RequestReresolutionLocked() override;
+ private:
+  class AresRequestWrapper : public InternallyRefCounted<AresRequestWrapper> {
```

Comment:
A couple of structural comments (I won't block the PR on this, just want to consider):

The thread safety around the combined request initialization and starting in `StartRequest` is somewhat subtle to reason about, since it relies on the `PollingResolver` to synchronize its access to the `Orphanable` returned from `StartRequest`.

What do you think about changing the `OrphanablePtr<Orphanable> StartRequest` API to `OrphanablePtr<PollingResolver::Request> CreateRequest`, where `PollingResolver::Request` looks like:

```
class Request : InternallyRefCounted<Request> {
  public:
    Start();
}
```

For c-ares, `PollingResolver::Request` would be implemented by `AresRequestWrapper` pretty much as is.

For the native resolver, we could introduce `NativeRequestWrapper` which would hold a `RefCountedPtr<DNSResolver::Request> request_;`, as well as `RefCountedPtr<PollingResolver>` (getting rid of the manual ref in the native resolver).

We could then go a step further and make `PollingResolver` a final class, and merge all of the remaining  `AresClientChannelDNSResolver` and `NativeClientChannelDNSResolver` methods into their corresponding `PollingResolver::Request` implementations. With `PollingResolver` e.g. taking a factory function in it's ctor for creating `PollingResolver::Request`.

Getting rid of the `AresClientChannelDNSResolver` and `NativeClientChannelDNSResolver` sub-class types might improve readability, since they are pretty thin now anyways.

Reply:
> The thread safety around the combined request initialization and starting in `StartRequest` is somewhat subtle to reason about, since it relies on the `PollingResolver` to synchronize its access to the `Orphanable` returned from `StartRequest`.

That's true, but `PollingResolver` is explicitly designed to make that guarantee.  In any polling resolver, the request is going to complete asynchronously, and when it does, we're going to have to hop back into the `WorkSerializer` to process the result.  `PollingResolver` handles that so that subclasses don't need to worry about it, so this access will always be safe.

Note that this is not intended to be a public API, so I'm fine with having it depend on the semantics of `WorkSerializer`.

> 
> What do you think about changing the `OrphanablePtr<Orphanable> StartRequest` API to `OrphanablePtr<PollingResolver::Request> CreateRequest`, where `PollingResolver::Request` looks like:
> 
> ```
> class Request : InternallyRefCounted<Request> {
>   public:
>     Start();
> }
> ```

I thought about that -- in fact, I had actually implemented something like that in an earlier draft of this PR -- but I decided against it, because I didn't want to force implementations to jump through hoops to avoid diamond dependency problems.

The native DNS resolver is a prime example of that: the iomgr DNS resolution API already returns a `DNSResolver::Request` object, which itself inherits from `InternallyRefCounted<>`.  The native DNS resolver can't return an object that inherits from both `DNSResolver::Request` *and* from `PollingResolver::Request` without hitting a diamond dependency problem, so it would be forced to create a new wrapper object that inherits from `PollingResolver::Request` and contains the `DNSResolver::Request` object that does all the real work.  That seems unnecessarily complex and inefficient, especially given that `PollingResolver` doesn't actually care whether the request object is ref-counted or not; all it really cares about is that it's orphanable, so that it can cancel it.

> 
> For c-ares, `PollingResolver::Request` would be implemented by `AresRequestWrapper` pretty much as is.

In the long run, I don't think we really want to have `AresRequestWrapper` in the first place.  If #25108 hadn't been reverted (and I would still like to see it rolled forward at some point), the c-ares resolver would have been able to just drop in its existing `AresRequest` object here.  But if we went with the `PollingResolver::Request` approach, the c-ares resolver would have had to jump through exactly the same hoops as the native DNS resolver to avoid the diamond dependency problem.

> 
> For the native resolver, we could introduce `NativeRequestWrapper` which would hold a `RefCountedPtr<DNSResolver::Request> request_;`, as well as `RefCountedPtr<PollingResolver>` (getting rid of the manual ref in the native resolver).

I agree that having a way to avoid dealing with this ref manually would be better.  But once we can use C++14, we can do that anyway by making the completion callback a lambda that captures the `RefCountedPtr<>` using move semantics.  (Even with C++11, we could actually use the lambda-capture approach instead of dealing with the refs manually, but that would result in an extra unnecessay ref/unref pair due to the lack of lambda move-capture semantics.)

> 
> We could then go a step further and make `PollingResolver` a final class, and merge all of the remaining `AresClientChannelDNSResolver` and `NativeClientChannelDNSResolver` methods into their corresponding `PollingResolver::Request` implementations. With `PollingResolver` e.g. taking a factory function in it's ctor for creating `PollingResolver::Request`.
> 
> Getting rid of the `AresClientChannelDNSResolver` and `NativeClientChannelDNSResolver` sub-class types might improve readability, since they are pretty thin now anyways.

I think that would force the interface for implementations to be a little too narrow, which would make it hard to expand it in the future if needed.

Also, it's worth noting that once AJ moves the c-ares code inside of the iomgr API, we will have only one client channel DNS resolver implementation anyway, so any small amount of duplication here will go away at that point anyway.

Reply:
Sounds good, thanks for context